### PR TITLE
feat/fix: currency column

### PR DESCRIPTION
**Pull request by mertmit**
## Change Summary

Currency locale and currency code can be customized.

Refs: #572 #1042 #1352 #1647

- [x] Currency Column
  - [x] Multiple Locale Support
Used Intl.NumberFormat to format and validate
  - [x] Multiple Currency Code Support (ISO 4217)
- [x] Validation
I propose changing validator to isFloat instead because with this implementation we handle currency part apart from the value
- [x] Migration
  - [x] Handle extra variables as colOptions instead of seperate meta columns
  - [x] Change validator for migrated columns
There is no need to migrate validators with latest implementation

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)
- [x] fix

## Test/ Verification

image
image

## Additional information

- Country code reference: https://en.wikipedia.org/wiki/ISO_4217

### CODE REVIEW

File: packages/nc-gui/components/project/spreadsheet/components/editColumn.vue [hunk 1]
```diff
@@ -597,6 +620,16 @@ export default {
         this.newColumn.dtxp = this.column.dtxp
       }
 
+      if (this.newColumn.uidt === 'Currency' && this.column?.uidt !== 'Currency') {
```
Comment (pranavxc) [inline packages/nc-gui/components/project/spreadsheet/components/editColumn.vue:623]: Use `UiTypes` enum for referring uidt whenever possible. So in the above case, you can replace `Currency` with `UiTypes.Currency`, where you can import `UiTypes` from `nocodb-sdk` package.

**Review by pranavxc — CHANGES_REQUESTED** [review]

**Review by wingkwong — COMMENTED** [review]
Just a heads up. We are putting v2 migration stuff in different places. Here's something you should change soon.

1. Move the changes you made in `XcMigrationSource.ts` to `XcMigrationSourceV2.ts`
2. Move `nc_012_add_currency_column.ts` to `migrationsv2` instead of `migrations`. You may also need to change the file name as `nc_012` is being occupied. 

However, we haven't merged it to `develop` branch yet. Will ping you when it is ready.

**Comment by mertmit** [comment]
I will prepare migrations according to your instructions and wait for your ping to push proper changes updated with latest develop branch.

Meanwhile I updated UI:
image

**Comment by wingkwong** [comment]
@mertmit Now it is in develop branch. Please pull the latest code and merge it accordingly. Thanks.

**Comment by mertmit** [comment]
**Caution:** The migration still misses changing validate column to 'isFloat' for existent currency columns.

**Comment by wingkwong** [comment]
@mertmit Marking it as Draft first. Please request a review again when it is ready.

**Comment by mertmit** [comment]
https://github.com/nocodb/nocodb/pull/1687/commits/202df2d731727e15e1a7e6c5cead7a233d37dd91
on this commit I've created a custom validators file as this might come in handy in future (if we want to bypass any function provided by validate or want to remap as in this case)

**Comment by wingkwong** [comment]
@mertmit Please rebase and resolve the conflicts first.

**Comment by o1lab** [comment]
Im wondering whether it would be good idea to not keep separate table for currency. That way we don't end up creating a lot of tables for each types. A simple json column or text column (stringify json) would be sufficient ? @wingkwong @pranavxc

File: packages/nocodb/src/lib/noco-models/Column.ts [hunk 2]
```diff
@@ -495,6 +510,18 @@ export default class Column<T = any> implements ColumnType {
       }
     }
 
+    {
+      let currs = await NocoCache.getList(CacheScope.COL_CURRENCY, [id]);
+      if (!currs.length) {
+        currs = await ncMeta.metaList2(null, null, MetaTable.COL_ROLLUP, {
```
Comment (wingkwong) [inline packages/nocodb/src/lib/noco-models/Column.ts:516]: I think it should be `COL_CURRENCY` instead of `COL_ROLLUP`

**Comment by wingkwong** [comment]
> Im wondering whether it would be good idea to not keep separate table for currency. That way we don't end up creating a lot of tables for each types. A simple json column or text column (stringify json) would be sufficient ? @wingkwong @pranavxc

@mertmit please correct me if I'm wrong.

I think the table is not for storing all the possible currency options but storing the meta currency (locale & code) that users choose. I think keeping it in a separate table would make it more clean.

**Comment by mertmit** [comment]
@wingkwong yes it only keeps colOptions (column_id, currency_locale, currency_code for currency) in created meta table.
@o1lab actually this is how we handle colOptions for other columns with colOptions. (formula, lookup etc.)

**Comment by wingkwong** [comment]
@mertmit Please rebase and request review again if it's ready. Thanks.

**Comment by mertmit** [comment]
@wingkwong ready for a review. Thanks.

**Comment by mertmit** [comment]
> https://github.com/nocodb/nocodb/pull/1687/commits/202df2d731727e15e1a7e6c5cead7a233d37dd91
> on this commit I've created a custom validators file as this might come in handy in future (if we want to bypass any function provided by validate or want to remap as in this case)

@wingkwong This note left from first state I reverted this change, now it is handled within custom validators which I stated in this comment. You can check the commit in this comment to review validation.

**Review by wingkwong — APPROVED** [review]

**Review by pranavxc — APPROVED** [review]

**Pull request merged**